src/Lib.P2P/Cryptography/EphermalKey.cs

```diff
@@ -0,0 +1,131 @@
+    ///   uncompressed ECPoint; the first byte s a 4 and followed by X and Y co-ordinates.
+    ///   <para>
+    ///   It as assummed that the curve name is known a priori.
+    ///   </para>
+    /// </remarks>
+    public sealed class EphermalKey
+    {
+        private ECPublicKeyParameters _publicKey;
+        private ECPrivateKeyParameters _privateKey;
+
+        /// <summary>
+        ///   Gets the IPFS encoding of the public key.
+        /// </summary>
+        /// <returns>
+        ///   Returns the uncompressed EC point.
+        /// </returns>
+        internal byte[] PublicKeyBytes() { return _publicKey.Q.GetEncoded(false); }
+
+        /// <summary>
+        ///   Create a shared secret between this key and another.
+        /// </summary>
+        /// <param name="other">
+        ///   Another ephermal key.
+        /// </param>
+        /// <returns>
+        ///   The shared secret as a byte array.
+        /// </returns>
+        /// <remarks>
+        ///   Uses the ECDH agreement algorithm to generate the shared secet.
+        /// </remarks>
+        public byte[] GenerateSharedSecret(EphermalKey other)
+        {
+            var agreement = AgreementUtilities.GetBasicAgreement("ECDH");
+            agreement.Init(_privateKey);
+            var secret = agreement.CalculateAgreement(other._publicKey);
+            return BigIntegers.AsUnsignedByteArray(agreement.GetFieldSize(), secret);
+        }
+
+        /// <summary>
+        ///   Create a public key from the IPFS ephermal encoding.
+        /// </summary>
+        /// <param name="curveName">
+        ///   The name of the curve, for example "P-256".
+        /// </param>
+        /// <param name="bytes">
+        ///   The IPFS encoded ephermal key.
+        /// </param>
+        internal static EphermalKey CreatePublicKeyFromDfs(string curveName, byte[] bytes)
+        {
+            var ecP = ECNamedCurveTable.GetByName(curveName);
+            if (ecP == null)
+            {
+                throw new KeyNotFoundException($"Unknown curve name '{curveName}'.");
+            }
+            
+            var domain = new ECDomainParameters(ecP.Curve, ecP.G, ecP.N, ecP.H, ecP.GetSeed());
+            var q = ecP.Curve.DecodePoint(bytes);
+            return new EphermalKey
+            {
+                _publicKey = new ECPublicKeyParameters(q, domain)
+            };
+        }
+
+        /// <summary>
+        ///   Create a new ephermal key on the curve.
+        /// </summary>
+        /// <param name="curveName">
+        ///   The name of the curve, for example "P-256".
+        /// </param>
+        /// <returns>
+        ///   The new created emphermal key.
+        /// </returns>
+        public static EphermalKey Generate(string curveName)
+        {
+            var ecP = ECNamedCurveTable.GetByName(curveName);
+            if (ecP == null)
+            {
+                throw new Exception($"Unknown curve name '{curveName}'.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)